### PR TITLE
C++ Framework and validation tests for operator validation and performance

### PR DESCRIPTION
**Pull request by cjolivier01**
## Description ##

C++ Framework and validation tests for operator validation and performance
Normally used for gtest tests.
Note: Sparse is not fully supported in this version.

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated.
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Intersting edge cases to note here

### CODE REVIEW

File: src/c_api/c_api_ndarray.cc [hunk 1]
```diff
@@ -35,59 +35,10 @@
 #include "./c_api_common.h"
 #include "../common/utils.h"
 #include "../common/exec_utils.h"
+#include "../imperative/imperative_utils.h"
 
 using namespace mxnet;
 
-nnvm::NodeAttrs ParseAttrs(const nnvm::Op *op,
```
Comment (cjolivier01) [inline src/c_api/c_api_ndarray.cc]: These are moved to imperative_utils.h.  I added documentation in the header.

File: tests/cpp/include/test_legacy_op.h [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+ * \brief operator unit test utility functions
+ * \author Chris Olivier
+ *
+ * These classes offer a framework for developing, testing and debugging operators
+ * in C++.  They work for both CPU and GPU modes, as well as offer a timing
+ * infrastructure in order to test inidividual operator performance.
+ *
+ * Operator data can be validated against general logic,
+ * stored scalar values (which can be generated by this code from an existing operator via
+ * BasicOperatorData::dumpC(), as well as against each other (ie check that
```
Comment (cjolivier01) [inline tests/cpp/include/test_legacy_op.h]: Much of this code was moved from test_op.h and then ported to use the new templating scheme.  The general flow remains similar.

**Comment by piiswrong** [comment]
How does this work? Looks like most modifications are in tests/

**Comment by cjolivier01** [comment]
Yes, it’s a staging system to performance test the core operators like elemwise_add() in C++. Previously it just ddnthenlegacy ops like ActivationOp. Haibin may use it for sparse (he would probably have to make modifications) and I want to use it to validate the OMP/non-OMP op and general tuning.
It basically infers inputs and outputs and builds backward pass, running one or both ways, without all of the engine noise muddling up vtune (for example). I also have an optimized version of SGMon I want to test.

**Comment by cjolivier01** [comment]
Sample of new use of the code changes is core_op_runner_test.cc

**Comment by piiswrong** [comment]
So this is just benchmarking, not the actual performance tuning part yet

**Comment by cjolivier01** [comment]
Right

**Comment by cjolivier01** [comment]
Is this the proper naming for these sorts of ops? elemwise_xxx, for example? As "core"?

**Pull request merged**